### PR TITLE
[3/6][kimi-deterministic] Enable the BF16 DeepGEMM path for DeepEP normal mode deterministically

### PR DESCRIPTION
**Pull request by zyzshishui**
> **Stacked series (3/6)**: stacked on `kimi-deterministic/2-batch-invariant-qkv-a-proj` (#30818), so the diff below includes 1-2/6. This PR's own changes are the last 2 commits (ending 5a3bd294e). Combined CI: #30822.

## Motivation

BF16 expert weights enter the DeepGEMM BF16 grouped GEMM only under DeepEP low_latency; deepep normal falls back to other kernels.

## Modifications

- Drop the `enable_low_latency()` condition in `ep_moe/layer.py` and `unquant.py`.
- Under deterministic inference, zero-init the pre-permute buffers in `moe_runner/deep_gemm.py` (`input_tensor`, `input_tensor_scale`, `m_indices`) so pad regions cannot leak nondeterministic values.

## Accuracy Tests

Kimi-K2-Instruct-BF16, 16 GPUs (TP16 / DP8-attention / EP16, fa4 + deep_gemm + deepep normal, radix on), 128 prompts (1k-16k) x 512 tokens, abs tol 0.0: replay, shuffled concurrency-32, bs1 vs bs128, cached-prefix extend vs cold prefill, decode logprobs vs prefill rescoring (65,536 tokens) — all bitwise identical.

Baselines from two independent server instances are bitwise-identical.

## Speed Tests and Profiling

Zero-init only under deterministic inference (three fills per MoE layer); default path unchanged.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :x: [Run #29979964397](https://github.com/sgl-project/sglang/actions/runs/29979964397)
Latest PR Test (Extra): :x: [Run #29979964294](https://github.com/sgl-project/sglang/actions/runs/29979964294)

### CODE REVIEW

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 1]
```diff
@@ -755,7 +758,11 @@ def pre_permute_deepep_normal_to_deep_gemm(
             device=hidden_states.device,
             dtype=torch.float32,
         )
+        if get_server_args().enable_deterministic_inference:
+            input_tensor_scale.zero_()
     m_indices = torch.empty(all_tokens, device=hidden_states.device, dtype=torch.int32)
+    if get_server_args().enable_deterministic_inference:
+        m_indices.zero_()
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: ~**[suggestion]** `m_indices.zero_()` is likely redundant for determinism. With DeepEP `expert_alignment=128` (when JIT DeepGEMM is on) and `_fwd_kernel_ep_scatter_1`, every slot in `m_indices` is written with the owning expert id for the full aligned `num_recv_tokens_per_expert` range. Pre-zeroing does not change post-scatter contents. Pad nondeterminism comes from `input_tensor` slots that `ep_scatter_2` never writes (alignment padding), not from leftover `m_indices` values. Zeroing to `0` also means “expert 0” rather than DeepGEMM’s usual invalid marker (`-1`) if any unwritten hole ever existed—though with 128 alignment that hole should not exist.~

~**Suggestion:** Drop `m_indices.zero_()` unless you have a measured determinism failure tied specifically to it; if keeping a defensive fill, prefer documenting why, or use `fill_(-1)` only if intentionally marking unwritten rows as skipped (would require changing scatter semantics to match).~

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 2]
```diff
@@ -755,7 +758,11 @@ def pre_permute_deepep_normal_to_deep_gemm(
             device=hidden_states.device,
             dtype=torch.float32,
         )
+        if get_server_args().enable_deterministic_inference:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[suggestion]** `input_tensor_scale.zero_()` only runs on the non-`DEEPGEMM_SCALE_UE8M0` branch (UE8M0 already uses `torch.zeros`). For the BF16 DeepGEMM path this PR enables, `ep_scatter` sets `is_fp8 = recv_x_scale is not None and recv_x.dtype != torch.bfloat16`, so scale is not written/consumed when dispatch dtype is bf16 and weights are bf16 (`_run_bf16_contiguous_gemm` ignores scales). Zeroing the scale buffer is therefore a no-op for the stated Kimi-K2 BF16 case; it only helps if FP8 activations share this pre-permute under deterministic mode.

**Suggestion:** Keep if you want shared FP8+deterministic coverage; otherwise gate scale zeroing on an FP8 path (or leave a one-line comment that it is for FP8 pad slots only). The buffer that actually matters for BF16 pad leakage is `input_tensor`.

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 3]
```diff
@@ -742,6 +743,8 @@ def pre_permute_deepep_normal_to_deep_gemm(
         device=hidden_states.device,
         dtype=hidden_states.dtype,
     )
+    if get_server_args().enable_deterministic_inference:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[nit]** `get_server_args().enable_deterministic_inference` is evaluated up to three times in one pre-permute. Flag is process-static, so this is not a correctness issue, only noise and three attribute lookups on a hot path when deterministic mode is on.

**Suggestion:** Cache once, e.g. `deterministic = get_server_args().enable_deterministic_inference`, then branch; optionally use `torch.zeros(...)` when true instead of `empty` + `zero_()` for clarity (same result).

**Review by ch-wan — APPROVED** [review]

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 4]
```diff
@@ -750,12 +759,12 @@ def pre_permute_deepep_normal_to_deep_gemm(
             dtype=torch.int,
         ).transpose(0, 1)
     else:
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: **[nit]** Comment correctly states scales only matter for FP8. Under det, the float-scale branch still uses `buffer_init` zeros for BF16 activations where the buffer is unused by scatter/GEMM. Not wrong; just extra fill on the BF16 det path this PR targets.

**Suggestion:** Optionally allocate/zero `input_tensor_scale` only when scales are used (FP8), or leave as-is given shared pre-permute.
